coverity issues

mpw96 opened the pull request:
Hi,
coverity scan at SAP found some issues. There are some more in scintilla but I figure they should be fixed there.

T&R
Marc-Philip

Commit:
fix possible overrun

Commit:
fix unitialized variable

Commit:
fix illegal memory access

Commit:
add missing zero padding

Commit:
avoid using invalid field

catalinr commented on src/common/strconv.cpp:
```diff
@@ -424,7 +424,7 @@ const wxCharBuffer wxMBConv::cWC2MB(const wchar_t *pwz) const
     {
         const size_t nLen = FromWChar(NULL, 0, pwz);
-        if ( nLen != wxCONV_FAILED )
+        if ( nLen != wxCONV_FAILED && nLen > 0 )
```
`nLen` is always positive. `>0` is not really needed.

catalinr commented on src/common/strvararg.cpp:
```diff
@@ -703,52 +703,55 @@ wxFormatString::ArgumentType DoGetArgumentType(const CharType *format,
-                 "requested argument not found - invalid format string?" );
-
-    switch ( parser.pspec[n-1]->m_type )
+    if (0<n)
```
Wouldn't this be better as a `wxCHECK_MSG` at the beginning of this function? It would avoid all the blank changes.
Apart from that:
- `n` is always positive, so comparing it with `0` in the condition is not needed;
- `if` spacing is improper, see the previous condition at line 694.

mpw96 replied:
inside this function you cannot be sure that n is always positive, because the format string might not contain arguments, and parser.nargs can become zero.

catalinr replied:
> inside this function you cannot be sure that n is always positive

You mean `unsigned n` ?

vadz replied:
I could be wrong, but I don't think `n` could be 0 here, it's the index of the argument and there must be the format string before. So IMO this change is really unneeded and it would be nice to find some way to pacify Coverity without introducing it.

Of course, disregard this if you can actually reproduce the problem in this code (how?).

catalinr commented on src/common/wxcrt.cpp:
```diff
@@ -685,7 +685,8 @@ int wxVsnprintf(wchar_t *str, size_t size, const wxString& format, va_list argpt
     // VsnprintfTestCase reveals that glibc's implementation of vswprintf
     // doesn't nul terminate on truncation.
-    str[size - 1] = 0;
+    if (0<size)
```
Unneeded comparison for unsigned type; incorrect spacing.

mpw96 replied:
size can be zero here.

catalinr replied:
`if ( size )` is enough.

catalinr commented on src/html/helpdata.cpp:
```diff
@@ -339,9 +339,14 @@ inline static void CacheWriteString(wxOutputStream *f, const wxString& str)
-    wxCharBuffer str(len-1);
-    f->Read(str.data(), len);
-    return wxString(str, wxConvUTF8);
+    if (0<len)
```
A `wxCHECK_MSG` might be better here.

mpw96 replied:
IMHO returning an empty string in that case is exactly what is expected, isn't it?

catalinr replied:
`wxCHECK_MSG` has `retValue`, and it can be `wxString()`, can't it?

vadz replied:
I don't think `wxCHECK` is appropriate, because condition failure doesn't indicate a programming error. But returning silently is wrong too, AFAICS from this (unfamiliar) code, size is really not supposed to be 0 here, so we should arguably log an error about the book being corrupted or something like that.

catalinr commented on src/stc/scintilla/src/PositionCache.cxx:
```diff
@@ -687,7 +687,12 @@ void PositionCache::MeasureWidths(Surface *surface, const ViewStyle &vstyle, uns
+			// If lenSegment is 0 here, we have an infinite loop, which is bad anyway.
+			// If startSegment is 0 here as well, we access positions at a bad position.
+			// So don't do that if both are 0.
+			if (startSegment + lenSegment > 0) {
```
Unneeded `> 0` part.

mpw96 replied:
if both startSegment and lenSegment are zero we have a problem, IMHO it's necessary.

catalinr replied:
`if (startSegment + lenSegment)` is enough.

catalinr replied:
p.s. This change should be proposed upstream, to scintilla.

lanurmi replied:
While `if (startSegment + lenSegment)` is enough, I think leaving out the `> 0` makes it less readable.

catalinr commented on include/wx/ctrlsub.h:
```diff
@@ -199,7 +199,11 @@ class WXDLLIMPEXP_CORE wxItemContainer : public wxItemContainerImmutable
+        void* clientDataArray[1];
+        clientDataArray[0] = clientData;
+        return AppendItems(item, clientDataArray);
+    }
```
Is this the only way to do it? Because it is pretty bad for readability..

vadz replied:
Also, what's the actual problem here? I don't really see one.

catalinr commented on src/jpeg/jctrans.c:
```diff
@@ -286,6 +286,8 @@ compress_output (j_compress_ptr cinfo, JSAMPIMAGE input_buf)
   jpeg_component_info *compptr;
 
+  memset(MCU_buffer, 0, C_MAX_BLOCKS_IN_MCU * sizeof(JBLOCKROW));
+
```
This change should be proposed upstream.

vadz commented on include/wx/private/wxprintf.h:
```diff
@@ -902,7 +902,7 @@ struct wxPrintfConvSpecParser
         // warn if we lost any arguments (the program probably will crash
         // anyhow because of stack corruption...)
-        if ( nargs == wxMAX_SVNPRINTF_ARGUMENTS )
+        if ( nargs >= wxMAX_SVNPRINTF_ARGUMENTS )
```
I don't think it's really safer, it just hides a potential bug in the code as `nargs` is never supposed to become strictly greater than the limit, so it makes things less clear.

What exactly is Coverity's problem here?

vadz commented:
Thanks for the fixes, it's nice to have them and at least one seems to be a real problem.

However it would be really helpful if you could use more verbose commit messages and it would also be great if you could mention Coverity errors fixed by each of them.

Other than that, I agree with @catalinr comments (thanks for the review!).

mpw96 commented:
Hi,
- I've fixed the formatting in the changes, where there was no discussion.
- I've removed the changes where there was discussion.
- I've removed the changes which should be fixed in upstream projects.

All changes except "that check is safer" were found by Coverity 8.5.

Regards
Marc-Philip

vadz commented:
Thanks, I'll merge this in as soon as the CI builds become green (not that I expect them to fail...).

vadz commented:
Just FYI: I've applied the same fixes to 3.0 branch as well as they seem to be necessary there too.

Thanks again!
